epuck_ros2_cpp/src/driver.cpp

```diff
@@ -0,0 +1,454 @@
+    msg.child_frame_id = "base_link";
+    msg.twist.twist.linear.x = v;
+    msg.twist.twist.linear.z = omega;
+    msg.pose.pose.position.x = position_x;
+    msg.pose.pose.position.y = position_y;
+    msg.pose.pose.orientation = euler_to_quaternion(0, 0, angle);
+    odometry_publisher->publish(msg);
+
+    // Pack & publish transforms
+    geometry_msgs::msg::TransformStamped tf;
+    tf.header.stamp = stamp;
+    tf.header.frame_id = "odom";
+    tf.child_frame_id = "base_link";
+    tf.transform.translation.x = position_x;
+    tf.transform.translation.y = position_y;
+    tf.transform.translation.z = 0.0;
+    tf.transform.rotation = euler_to_quaternion(0, 0, angle);
+    dynamic_broadcaster->sendTransform(tf);
+  }
+
+  void update_callback() {
+    int success;
+    int retry_count;
+    rclcpp::Time stamp;
+
+    // I2C: Set address
+    retry_count = 1;
+    success = 0;
+    while (!success && retry_count > 0) {
+      success = i2c_main->set_address(0x1F);
+      retry_count--;
+    }
+
+    // I2C: Write/Read
+    retry_count = 3;
+    success = 0;
+    while (!success && retry_count > 0) {
+      // Write
+      msg_actuators[MSG_ACTUATORS_SIZE - 1] = 0;
+      for (int i = 0; i < MSG_ACTUATORS_SIZE - 1; i++)
+        msg_actuators[MSG_ACTUATORS_SIZE - 1] ^= msg_actuators[i];
+      success = (i2c_main->write_data(msg_actuators, MSG_ACTUATORS_SIZE) == MSG_ACTUATORS_SIZE);
+
+      // Read
+      success &= (i2c_main->read_data(msg_sensors, MSG_SENSORS_SIZE) == MSG_SENSORS_SIZE);
+      char checksum = 0;
+      for (int i = 0; i < MSG_SENSORS_SIZE - 1; i++)
+        checksum ^= msg_sensors[i];
+      success &= (checksum == msg_sensors[MSG_SENSORS_SIZE - 1]);
+
+      retry_count--;
+    }
+
+    stamp = now();
+
+    if (success) {
+      publish_distance_data(stamp);
+      publish_odometry_data(stamp);
+      i2c_main_err_cnt = 0;
+    } else {
+      i2c_main_err_cnt++;
+    }
+  }
+
+  OnSetParametersCallbackHandle::SharedPtr callback_handler;
+
+  rclcpp::TimerBase::SharedPtr timer;
+  rclcpp::Publisher<sensor_msgs::msg::LaserScan>::SharedPtr laser_publisher;
+  rclcpp::Publisher<nav_msgs::msg::Odometry>::SharedPtr odometry_publisher;
+  rclcpp::Publisher<sensor_msgs::msg::Range>::SharedPtr range_publisher[9];
+  rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr subscription;
+
+  std::unique_ptr<tf2_ros::StaticTransformBroadcaster> laser_broadcaster;
+  std::unique_ptr<tf2_ros::TransformBroadcaster> dynamic_broadcaster;
+  std::unique_ptr<tf2_ros::StaticTransformBroadcaster> infrared_broadcasters[9];
+
+  std::unique_ptr<I2CWrapper> i2c_main;
+
+  int fh;
+  char msg_actuators[MSG_ACTUATORS_SIZE];
+  char msg_sensors[MSG_SENSORS_SIZE];
+
+  float prev_angle;
+  float prev_position_x;
+  float prev_position_y;
+  int16_t prev_left_wheel_raw;
+  int16_t prev_right_wheel_raw;
+  int odom_left_overflow;
+  int odom_right_overflow;
+  int i2c_main_err_cnt;
+
+  float wheel_distance;
+  float wheel_radius;
```

Comment:
I know it is a bit lat, but we usually name member variable with a `m` prefix and use camel-case, this way it is very simple to see in the code which variables are member variables and which are not.

Reply:
Camel-case notation implemented to all C++ files in `epuck_ros2_cpp` and `epuck_ros2_camera`

Reply:
Thank you